src/transformers/adapters/peft_mixin.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+from typing import Optional
+
+from ..utils import ADAPTER_CONFIG_NAME, cached_file, logging, requires_backends
+
+
+logger = logging.get_logger(__name__)
+
+
+class PeftAdapterMixin:
+    """
+    A class containing all functions for loading and using adapters weights that are supported in PEFT library.
+    Currently supported PEFT methods are all non-prefix tuning methods
+    """
+
+    def load_adapter(
+        self,
+        peft_model_id: str,
+        adapter_name: Optional[str] = "default",
+        revision: Optional[str] = None,
+        use_auth_token: Optional[str] = None,
+        commit_hash: Optional[str] = None,
+    ):
+        """
+        Load adapter weights from file. Requires peft as a backend to load the adapter weights
+        """
+        requires_backends(self.load_adapter, "peft")
+
+        from peft import LoraConfig, PeftModel, create_and_replace
+        from peft.utils import set_peft_model_state_dict
+        from peft.utils.other import TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING
+
+        self.peft_config = {}
+        self.modules_to_save = None
+
+        adapter_config_file = self._find_adapter_config_file(
+            peft_model_id,
+            revision=revision,
+            use_auth_token=use_auth_token,
+            commit_hash=commit_hash,
+        )
+
+        if adapter_config_file is None:
+            raise ValueError(
+                f"adapter model file not found in {peft_model_id}. Make sure you are passing the correct path to the "
+                "adapter model."
+            )
+
+        # TODO: automatically infer the correct config class
+        loaded_peft_config = LoraConfig.from_pretrained(
+            peft_model_id,
+            revision=revision,
+            use_auth_token=use_auth_token,
+            commit_hash=commit_hash,
+        )
+
+        if not hasattr(loaded_peft_config, "target_modules"):
+            target_modules = TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING[self.config.model_type]
+            loaded_peft_config.target_modules = target_modules
+
+        # TODO: constraint this to single adapter
+        if adapter_name not in self.peft_config:
+            self.peft_config[adapter_name] = loaded_peft_config
+        else:
```

Comment:
Can the `else` condition ever be met? `self.peft_config` is initialized as `{}` above. IIUC, right now, if a user calls `load_adapter` twice, the 2nd adapter will just override the 1st adapter, whereas the intention seems to be that the 2nd call should fail. Is that right?

Reply:
Ah yes correct, will update accordingly

Reply:
Now this is not needed anymore and is dealt directly on PEFT side, I found that more intuitive, please check this specific commit: https://github.com/huggingface/peft/pull/749/commits/8053c88bbdd1651cd75b59b6cb0111cdb1ddd41b